File: typescript/patterns-use-cases/src/sagas/clients/hotel_client.ts
import { TerminalError } from "@restatedev/restate-sdk";

export const hotelClient = {
  book: async (
    customerId: string,
    _req: { arrivalDate: string; departureDate: string }
  ) => {
    console.error("[👻 SIMULATED] This hotel is fully booked!");
    throw new TerminalError("[👻 SIMULATED] This hotel is fully booked!");
  },
  cancel: async (customerId: string) => {
    console.info(`Hotel cancelled for customer ${customerId}`);
  },
};

File: python/patterns-use-cases/sagas/clients/hotel_client.py
import logging

from pydantic import BaseModel
from restate.exceptions import TerminalError

logger = logging.getLogger(__name__)


class HotelRequest(BaseModel):
    arrival_date: str
    departure_date: str


async def book(customer_id: str, hotel: HotelRequest):
    # this should implement the communication with the hotel
    # provider's APIs
    logger.error("[👻 SIMULATED] This hotel is fully booked!")
    raise TerminalError("[👻 SIMULATED] This hotel is fully booked!")


async def cancel(customer_id: str):
    # this should implement the communication with the hotel
    # provider's APIs
    logging.info(f"Hotel reservation cancelled for customer: {customer_id}")
